Integrate Nix with IPFS (#133)
NixOS/rfcs: rfcs/0133-git-hashing.md

1. Per the abstract vs concrete model of the Nix store in https://github.com/NixOS/nix/pull/6877, everything we are doing is simply flushing out alternative interpretations of the abstract model.
   This is the sense in which we are, per the Scheme mantra, "removing the weaknesses and restrictions that make additional features appear necessary":
   Instead of extending the model with new features, we are relaxing concrete model assumptions (e.g. references are always opaque store paths) while keeping the abstract model the same.

2. We also support plans to decouple the layers of Nix further, and update our educational and marketing material to reflect it.
   With Flakes and other post-2.0 features, the upper layers of Nix have gained an enormous amount of flexibility and sophistication.
   RFCs like this show that the so-far more sleepy lower layers also have plenty of potential to gain sophistication too.

   Embracing layering on technical, educational, communications, and managerial levels can scale our capacity to manage complexity and sophistication without the project growing out of control.

>    It will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.

Author: ```suggestion
   Layering will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.
```

Currently distributing Nix binaries takes a lot of bandwidth and storage.
This is a barrier to being a Nix user in areas of slower internet --- which includes the vast majority of the world's population at this time.
This is also a barrier to users running their own caches.

Content-addressing opens up a *huge* design space of solutions to get around such problems.

The first steps proposed below do *not* tackle this problem directly, but it lays the ground-work for future experiments in this direction.

## Source distribution and archival

Source code used by Nix expressions frequently goes off-line. It would be beneficial if there was some resistance to this form of bitrot.
The Software Heritage archive stores much of the source code that Nix expressions use. They would be a natural partner in this effort.

Unfortunately, as https://www.tweag.io/blog/2020-06-18-software-heritage/ describes at the end, a major challenge is the way Nix content-addresses software.
First of all, Nix hashes sources in bespoke ways that no other project will adopt.
Second of all, hashing tarballs instead of the underlying files leads non-normative details (compression, odd perms, etc.).

> We should natively support git file hashing, which Git repos and Software Heritage both support.

Reviewer: ```suggestion
We should natively support Git file hashing, which is supported both by Git repos and Software Heritage.
```

This will completely obliterate these issues.

Overall, we are building out a uniform way to work with source code, regardless of its origins or the exact tools involved.

# Detailed design
[design]: #detailed-design

Each item can be done separately provided its dependent items are also done.
These are the items we wish to commit to at this time.

> (The goals mentioned future work are, in a separate document, also broken down into a dependency graph of smaller steps.)

Author: ```suggestion
(The goals mentioned under [future work](#future-work) are, in a separate document, also broken down into a dependency graph of smaller steps.)
```

 - Many store interface methods that today take store paths will need to also accept names & content address pairs.

   For stores that are purpose-built for Nix, like the ones we support today, all addressing can be done with store paths, so the current interface is fine.
   But for Nix-agnostic stores, store paths are rather useless as a key type because Nix-agnostic tools don't know about them.
   Those store can, however, understand content addresses.
   And from such a name + content address, we can always produce a store path again, so there is no loss of functionality with existing stores.

- Relax `ValidPathInfo` to merely require that *either* the pair of `NarHash` and `NarSize` or just `CA` alone be defined.

  As described in the first step, currently `NarHash` and `NarSize` are the *normative* fields which are used to verify a store object.
  But if the store object is content-addressed, we don't need these, because the content address (`CA` field) will also suffice, all by itself.
  
  Existing Nix stores types are still required to contain a `NarHash` and `NarSize`, which is good for backwards compat and don't come with a cost.

>   Only new nix-agnostic store types would take advantage of these new, relaxed rules.

Reviewer: ```suggestion
  Only new Nix-agnostic store types would take advantage of these new, relaxed rules.
```

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

We encourage anyone interested to check our tutorial in https://github.com/obsidiansystems/ipfs-nix-guide/ which demonstrates the above functionality.
Note at the time of writing this guide uses our original 2020 fork of Nix.

# Drawbacks
[drawbacks]: #drawbacks

## Complexity

The main cost is more complexity to the store layer.
For a few reasons we think this is not so bad.

Most importantly is the division of the work into a dependency graph of steps.
This allows us to slowly try out things like IPFS that leverage Git hashing, and not commit to more change than we want to up front.

Even if we do end up adopting everything though, we think for the following two reasons the complexity can still be kept manageable:

> 1. Per the abstract vs concrete model of the nix store in https://github.com/NixOS/nix/pull/6877 , everything we are doing is simply flushing out alternative interpretations of the abstract model.

Reviewer: ```suggestion
1. Per the abstract vs concrete model of the Nix store in https://github.com/NixOS/nix/pull/6877, everything we are doing is simply flushing out alternative interpretations of the abstract model.
```

>    This is the sense in which we are "removing the weaknesses and restrictions that make additional features appear necessary" per the Scheme mantra cited above:

Reviewer: ```suggestion
   This is the sense in which we are, per the Scheme mantra, "removing the weaknesses and restrictions that make additional features appear necessary":
```

It doesn't seem to be mentioned anywhere else.

Author: Guess it got removed!

   Instead of extending the model with new features, we are relaxing concrete model assumptions (e.g. references are always opaque store paths) while keeping the abstract model the same.

2. We also support plans to decouple the layers of Nix further, and update our educational and marketing material to reflect it.

>    With Flakes and other post-2.0 features, the upper layers of Nix have gained an enormous amount of flexibility and sophistication.
>    RFCs like this show that the so-far more sleepy lower layers also have plenty of potential to gain sophistication too.
> 
>    Embracing layering on technical, educational, communications, and managerial levels can scale our capacity to manage complexity and sophistication without the project growing out of control.

Reviewer: ```suggestion
```

This doesn't seem helpful to convey the argument.

   It will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.

>    That frees up "complexity budget" for project like this.

Reviewer: ```suggestion
   That frees up "complexity budget" for projects like this.
```

> 
>    We plan on more formally proposing this next.

Reviewer: ```suggestion
```

This seems unrelated and is unspecific.

## Git and Nix's file system data models do not entirely coincide

Nix puts the permission info of a file (executable bit for now) with that file, whereas Git puts it with the name and hash in the directory.
The practical effect of this discrepancy is that a root file (as opposed to directory) in Nix has permission info, but does not in Git.

If we are trying to convert existing Nix data into Git, this is a problem.
Assuming we treat "no permission bits" as meaning "non-executable", we will have a partial conversion that will fail on executable files without a parent directory.
Tricks like always wrapping everything in a directory get around this, but then we have to be careful the directory is exactly as expected when "unwrapping" in the other direction.

For now, we only focus on ingesting data *from* Git *to* Nix, and this side-steps the issue.

> That conversation is total (though not surjective), and so there is no problem for now.

Reviewer 2: ```suggestion
That mapping is total, i.e. all Git data can be mapped, and injective, i.e. each Git data has a unique Nix data representative (though not surjective, i.e. not all Nix data can be represented as a piece of Git data), and so there is no problem for now.
```

(saying total and "not surjective" is confusing to me, I have no strong opinion on keeping conversion, but using function vocabulary with conversion is a bit strange.)

feel free to drop the `i.e.` clarifications

Author: Ah yes this a great, and I'll keep the "i.e." for those that aren't familiar with the math.

shepherd-leader: amjoseph-nixpkgs
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Integrate Git hashing with Nix.

Nix should support content-addressed store objects using git blob + tree hashing, and Nix-unaware remote stores that serve git objects.

This follows the work done and described in https://github.com/obsidiansystems/ipfs-nix-guide/ .

# Motivation
[motivation]: #motivation

## Binary distribution

Currently distributing Nix binaries takes a lot of bandwidth and storage.
This is a barrier to being a Nix user in areas of slower internet --- which includes the vast majority of the world's population at this time.
This is also a barrier to users running their own caches.

Content-addressing opens up a *huge* design space of solutions to get around such problems.

The first steps proposed below do *not* tackle this problem directly, but it lays the ground-work for future experiments in this direction.

## Source distribution and archival

A goal of the Nix ecosystem is to package software in a way that never bitrots.
Getting in the way of that, however, is the fact source code frequently goes off-line.

> The Software Heritage archive is the best in the world, and a natural partner in this effort.

Author: Sure, sounds good

Unfortunately, as https://www.tweag.io/blog/2020-06-18-software-heritage/ describes at the end, a major challenge is the way nix content-addresses software.
First of all, Nix hashes sources in bespoke ways that no other project will adopt.

> Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).

Reviewer 3: ```suggestion
Second of all, using tarballs instead of the underlying files leaks non-normative details (compression, odd perms, etc.).
```
otherwise it should be "leak" ;)

> Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).

Author: ```suggestion
Second of all, hashing tarballs instead of the underlying files leads non-normative details (compression, odd perms, etc.).
```

> Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).

Author: Yes!

We should natively support git file hashing, which Git repos and Software Heritage both support.
This will completely obliterate these issues.

> Overall, are building out a uniform way to work with source code, regardless of its origins or the exact tools involved.

Author: ```suggestion
Overall, we are building out a uniform way to work with source code, regardless of its origins or the exact tools involved.
```

> ## Build adoption through seamless interop

Author: I am fine just cutting it entirely if people think the other motivation is sufficient.

> ## Build adoption through seamless interop

Reviewer 4: I don't mind the paragraph but I don't think it is critical to the proposal either.

> ## Build adoption through seamless interop

Author: It's gone. I can just blog about it :D

We don't want to "pander" in ways that will make Nix faddish but ultimately undermine it's popularity over the long haul (see Docker the company's woes).

One way to get around this tension to me is rather than pushing Nix towards the rest of the world, pushing the rest of the world towards us.
Like-minded projects emphasizing content-addressing are our *natural* partners, and we should work with them to promote Nix-*agnostic* standards that further our values and mission.

# Detailed design
[design]: #detailed-design

Each item can be done separately provided its dependent items are also done.
These are the items we wish to commit to at this time.
(See [Future Work](#future-work) for other steps left for later.)

## Git file hashing

- **Purpose**: Source distribution and archival

In addition to the various forms of content-addressing Nix supports today ("text", "fixed" with either "flat" or "nar" serialization of file system objects), Nix should support Git hashing.
This support entails two basic things:

 - Content addresses are used to compute store paths.
 - Content addresses are used to verify store object integrity.

> Git hashing would not support references (since references in Nix's sense are not a Git concept), but that is not an issue for the intended use-case of exchanging source code.

Reviewer 3: I don't think supporting fewer formats overall is desirable, and find it hard to imagine a future where NAR goes away -- Nix's astounding capabilities for running software from the past would suffer significantly from such a deprecation. I'm more in favour of supporting _more_ formats :)

As for extending Git trees, I think representing it in-band as you suggest would open a lot of potential for breakage, and a new format that builds on the strengths of Git's representation while avoiding its weaknesses -- as [Tvix is doing, with a conscious and justified rejection of Git trees](https://code.tvl.fyi/tree/tvix%2Fstore%2Fdocs%2Fwhy-not-git-trees.md) -- is probably a better path forwards.

> Git hashing would not support references (since references in Nix's sense are not a Git concept), but that is not an issue for the intended use-case of exchanging source code.

Author: ```suggestion
Git hashing would not (in this first proposed version) support references, since references in Nix's sense are not part of Git's data model.
This is OK for now; encoding references is not needed for the intended initial use-case of exchanging source code.
```

## Git file hashing for `buitins.fetch*`

- **Purpose**: Source distribution and archival

> - **Depends on**: Git file hashing,

Reviewer 3: A visual representation of this graph would be very helpful for understanding the way that the parts interact with each other :)

- **Purpose**: Source distribution and archival
- **Depends on**: Content address or store path in Store interface,

As described in the first step, currently `NarHash` and `NarSize` are the *normative* fields which are used to verify a store object.
But if the store object is content-addressed, we don't need these, because the content address (`CA` field) will also suffice, all by itself.
Relax the abstract `ValidPathInfo` type to merely require that *either* the pair of `NarHash` and `NarSize` or just `CA` alone be defined.

Existing Nix stores types are still required to contain a `NarHash` and `NarSize`, which is good for backwards compat and don't come with a cost.
Only new nix-agnostic store types would take advantage of these new, relaxed rules.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

We encourage anyone interested to check our tutorial in https://github.com/obsidiansystems/ipfs-nix-guide/ which demonstrates the above functionality.
Note at the time of writing this guide uses our original 2020 fork of Nix.

# Drawbacks
[drawbacks]: #drawbacks

## Complexity

The main cost is more complexity to the store layer.

> For a few reason we think this is not so bad.

Reviewer 4: ```suggestion
For a few reasons we think this is not so bad.
```

Most importantly is the division of the work into a dependency graph of steps.

> This allows us to slowly try out things like IPFS that leverage git hashing, and not commit to more change than we want to up front.

Reviewer 4: ```suggestion
This allows us to slowly try out things like IPFS that leverage Git hashing, and not commit to more change than we want to up front.
```

   With Flakes and other post-2.0 features, the upper layer of Nix have gained an enormous amount of flexibility and sophistication.
   RFCs like this show that the so-far more sleepy lower layers also have plenty of potential to gain sophistication too.

   Embracing layering on technical, educational, communications, and managerial levels can scale our capacity to manage complexity and sophistication without the project growing out of control.
   It will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.
   That frees up "complexity budget" for project like this.

   We plan on more formally proposing this next.

## Git and Nix's file system data models do not entirely coincide

Nix puts the permission info of a file (executable bit for now) with that file, whereas Git puts it with the name and hash in the directory.
The practical effect of this discrepancy is that a root file (as opposed to directory) in Nix has permission info, but does not in Git.

If we are trying to convert existing Nix data into Git, this is a problem.

> Assuming we treat "no permission bits" as meaning "non-executable", we will have a partial conversion that will fail on executable bare files.

Reviewer 3: ```suggestion
Assuming we treat "no permission bits" as meaning "non-executable", we will have a partial conversion that will fail on executable files without a parent directory.
```

But does the Git model even need to support files without a parent directory? I can see some value in allowing content-addressing models that don't support all things that can live in a store.

That conversation is total (though not surjective), and so there is no problem for now.

# Alternatives
[alternatives]: #alternatives

The dependency graph of steps can be sliced to save some for future work.
For now they are all written together, but during the RFC meetings we will decide which steps (if any) to ratify now, and which steps to save for later.

# Unresolved questions
[unresolved]: #unresolved-questions

None at this time.

# Future work --- IPFS
[future]: #future-work

## Motivation

### Binary Distribution

IPFS is a potential solution leveraging content addressing for a P2P CDN.
It out-of-the-box match's Nix's data model extremely well.

### Source distribution and archival

IPFS also supports git hashing, and so we also provide a good way for people and institutions to "pin" the sources they need, especially if those sources include private ones SWH won't have.
Finally, per [Obsidian's bridging work](https://github.com/obsidiansystems/go-ipfs-swh-plugin), we have a 3 way integration between IPFS, SWH, and Nix.
Data can be directly downloaded from SWH via HTTPS, or indirectly via IPFS, which can act as a CDN to not put as much load on SWH's servers.

> ### Not just IPFS

Reviewer 4: This may be too far off topic but since IPFS was mentioned I recall that we discussed that IPFS has a max block size whereas Git doesn't. Can you elaborate how this will affect our Git hashing implementation and eventually our IPFS implementation? It seems that IPFS is one of the major motivators of this work so it would be nice to have a rough idea of how this will work in the end.

Author: It won't work for larger files until that restriction is lifted, but there is a few ideas on how to do that.

Author: Collapsing IPFS threads. I hope even if we just were to do git hashing and not IPFS the normative part would still stand on it's own so don't need to resolve this completely now?

---
feature: git-hashing

> start-date: (fill me in with today's date, YYYY-MM-DD)

Reviewer 5: Missing date

Author: ```suggestion
start-date: 2022-08-27
```

Author: Thanks

Note at the time of writing this guide uses our original 2020 fork of Nix.

# Drawbacks
[drawbacks]: #drawbacks

## Complexity

The main cost is more complexity to the store layer.
For a few reason we think this is not so bad.

Most importantly is the division of the work into a dependency graph of steps.
This allows us to slowly try out things like IPFS that leverage git hashing, and not commit to more change than we want to up front.

Even if we do end up adopting everything though, we think for the following two reasons the complexity can still be kept manageable:

2. Per the abstract vs concrete model of the nix store in https://github.com/NixOS/nix/pull/6877 , everything we are doing is simply flushing out alternative interpretations of the abstract model.
   This is the sense in which we are "removing the weaknesses and restrictions that make additional features appear necessary" per the Scheme mantra cited above:
   Instead of extending the model with new features, we are relaxing concrete model assumptions (e.g. references are always opaque store paths) while keeping the abstract model the same.

> 3. We also support plans to decouple the layers of Nix further, and update our educational and marketing material to reflect it.

Reviewer 6: ```suggestion
2. We also support plans to decouple the layers of Nix further, and update our educational and marketing material to reflect it.
```

Github's markdown renderer doesn't seem to be renumbering these for some reason.

>    With Flakes and other post-2.0 features, the upper layer of Nix have gained an enormous amount of flexibility and sophistication.

Reviewer 6: ```suggestion
   With Flakes and other post-2.0 features, the upper layers of Nix have gained an enormous amount of flexibility and sophistication.
```

For now they are all written together, but during the RFC meetings we will decide which steps (if any) to ratify now, and which steps to save for later.

# Unresolved questions
[unresolved]: #unresolved-questions

None at this time.

# Future work --- Software Heritage
[future]: #future-work-swh

## Motivation

Software Heritage is a great way to get source code that is no longer found at its original location online.
It would be really great to use it as a substituter to fetch source code that fails to be fetched the default way.

## Detailed Designed

- **Depends on**: Nix-agnostic content-addressing "stores"

[`GET /api/1/raw/(swhid)/`](https://docs.softwareheritage.org/devel/swh-web/uri-scheme-api.html#get--api-1-raw-(swhid)-)
is a relatively new Software Heritage API endpoint to get a git object for the given SWHID.
We can used this to write a Software Heritage store, which will accomplish the task layed out in the previous motivation section.

# Future work --- IPFS
[future]: #future-work-ipfs

## Motivation

### Binary Distribution

IPFS is a potential solution leveraging content addressing for a P2P CDN.

> It out-of-the-box match's Nix's data model extremely well.

Reviewer 4: This sentence doesn't read properly for me. Did you mean to say "It's out-of-the box data model matches Nix's data model very well."?

Author: I think I meant
```suggestion
It, out-of-the-box, matches Nix's data model extremely well.
```

Reviewer 4: Ok that works for me. However I think that it is easier to read like this

```suggestion
Out-of-the-box it matches Nix's data model extremely well.
```

Author: Sure, though I think I'll put back in the single comma :)

### Source distribution and archival

IPFS also supports git hashing, and so we also provide a good way for people and institutions to "pin" the sources they need, especially if those sources include private ones Software Heritage won't have.
Finally, per [this IPFS plugin](https://blog.obsidian.systems/software-heritage-bridge/), we can have a 3 way integration between IPFS, Software Heritage, and Nix.
Data can be directly downloaded from Software Heritage via HTTPS, or indirectly via IPFS.

This would hopefully supercede the direct Software Heritage store outlined above.
This is because Software Heritage is primarily an archive, and we should be careful not to waste their limitted resources with copious egress bandwidth.
IPFS can act as a CDN to not put as much load on Software Heritage's servers.

> ### Not just IPFS

Reviewer 7: I initially didn't see the other marked-as-resolved threads, but I think this should be reworded to focus on the content-adressability, and marking IPFS as one possible vehicle/transport for that data.

Author: We perhaps should just delete all this text as move it to a draft new RFC :)

(See [Detailed Design](#detailed-design) for other steps left for later.)

### Augmented `narinfo`

- **Purpose**: Binary distribution

*This is taken from [RFC PR 122](https://github.com/lucasew/rfcs/blob/binary-cache-ipfs/rfcs/0122-binary-cache-ipfs.md), which was abandoned by its author.*

The purpose of this is a "hybrid" store where the narinfo metadata is still served via HTTPS, but the data itself is served via IPFS.

Today, a narinfo looks like this:

```
StorePath: /nix/store/gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
URL: nar/0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb.nar.xz
Compression: xz
FileHash: sha256:0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb
FileSize: 40360
NarHash: sha256:1ddv0iqq47j0awyw7a8dmm8bz71c6ifrliq53kmmsfzjxf3rwvb8
NarSize: 197528
References: 7gx4kiv5m0i7d7qkixq2cwzbr10lvxwc-glibc-2.27 gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
Deriver: 5sj6fdfym58sdaf3r5p87v4l8sj2zlvn-hello-2.10.drv
Sig: cache.nixos.org-1:K0thQEG60rzAK8ZS9f1whb7eRlIshlMDJAm7xvX1oF284H+PTqlicv/wGW6BIj+wWWONHvUZ2MYc+KDArekjDA==
```

This RFC proposes new key-value pairs that in this example would be:

```

> IpfsCid: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN

Reviewer 7: It's not entirely clear to me why we're using IPFS CIDs (and unixfs) here, instad of let's say, pointing to a raw git (blob or tree) object. This feels like a more logical choice, considering the first part starts with git tree hashing.

Can you make it more clear why unixfs was chosen?

In addition to that, I'd personally be in favor having `IpfsCid / IpfsEncoding` be one single field, and having specified in that RFC already how additional representations are handled. (Can these fields be repeated, or is this a ?-separated string of multiple possible representations?)

This allows us gradually introducing more representations, and pointing to more than one representation, without worrying about older Nix versions not supporting these protocols to not be able to parse it at all).

Reviewer 4: >  I'd personally be in favor having IpfsCid / IpfsEncoding be one single field

I think it is best to keep it separated. For the common operation of retrieving the file you only need the CID. The encoding settings are only needed if you have the store path and want to add it to IPFS to share with others which is probably a relatively rare operation.

Reviewer 4: Actually on further review it is somewhat mixed. `"method":"wrapped-nar"` is needed for both reading and resharing. Maybe it makes sense to move that property to `IpfsCid` (which will then need to be renamed) so that all reading properties are in the same place.

Reviewer 7: As written above, we might need to specify how multiple different encodings are specified, and thinking about "describing the NAR file twice, but with different chunking parameters" would be one usecase where it'd be quite complicated if cid and "generation method/interpretation" were two separate fields.

Reviewer 4: Would we ever want to describe the same NAR file twice with different chunking parameters? It seems that in general the creator should pick the parameters and to ensure maximum sharing everyone should re-use them.

In theory we don't ever need chunking parameters. Ideally the producer would share on IPFS and everyone would just download that CID. The main reason that we provide them is to help bootstrapping. For example right now https://cache.nixos.org isn't going to publish via IPFS, so the standardized chunking parameters will allow the community to do that. If the producer does publish the CID themselves then everyone would just download that and now worry about how it was chunked.

However interpretation is important as once you download the CID you need to know how to use it. It is needed for both reading and writing.

I agree that **if** we want to support multiple CIDs then it probably makes sense to group everything together to avoid parallel corresponding lists. However if we don't want to support mutliple CIDs then I think it makes sense to separate "things required to download and use" from "things required if you want to start seeding the CID".

Author: I think "This RFC" is misleading, because this is the the non-normative future work section, no?

IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}
```

Just as today, the `NarHash` and `NarSize` remain the *normative* way to lock down the store object the `narinfo` file describes.
Conversely, The `URL`, `FileHash` and `FileSize` by contrast are *informational*, describing not what the store object *is*, but *how to get it*.

The `IpfsCid` and `IpfsEncoding` are likewise informational, describing how to get the store object:

- `IpfsCid`: Native content address for IPFS.

- `IpfsEncoding`: Enough info to deterministically rebuild the IPFS representation from a non-IPFS copy of the store object.

   For now, `IpfsEncoding` will only support `unixfs-nar`, which works as follows:

   The NAR is itself wrapped in IPFS's [UnixFS](https://github.com/ipfs/specs/blob/main/UNIXFS.md).

>    This other format can be extracted from the CID (which is conceptually a pair of encoding metadata and a hash).

Reviewer 7: It's not clear to me what "wrapping" means here. Do you /decompose/ the NAR file, and use unixfs for its "seekability", or do you put the NAR file as a single file inside a unixfs? If the latter is the case, why bother wrapping unixfs around it at all (just to workaround the blocksize problem)?, and not just querying the NAR as a sha256 blob?

Reviewer 4: I believe that this is answered here: https://github.com/NixOS/rfcs/pull/133/files/165979cbedf6bff63dbf2379d61c8b38b631d63e#diff-4d468043c922b901f9d6c43b21703e397e19abdca0cb219ffb9e9005d9b9e5c2R278-R279

   For now, only IPFS's "unixfs" is supported.

>    `chunking`, `layout`, and `max-width` are tuning parameters for unixfs [described in the UnixFS spec](https://github.com/ipfs/specs/blob/main/UNIXFS.md#importing).

Reviewer 4: For now do we want to fix the supported parameters to simplify implementation? I think it may make sense to keep the format as a JSON blob as above but state that at this time only exactly that JSON string is supported. (Although readers are encouraged to support more flexibility in parameters).

This way we have a simple and well defined set of parameters to start. We can always expand and add more as we are sure that relevant clients support them.

Reviewer 7: I added another comment above - this RFC should at least specify how alternative/additional "not-yet-understood" representations would look like, to ensure we can add more of them without breaking code only implementing this RFC.

Author: This is all non-normative future work right? At least I thought it was.

Author: The no-yet-understood content addressing method boils down to unrecognized CA fields `ValidPathInfo` (narinfo, DB, etc.). But we can call that out explicitly.
